test/libraries/Safe150Migration.spec.ts

```diff
@@ -0,0 +1,771 @@
+import { expect } from "chai";
+import hre, { ethers, deployments } from "hardhat";
+import { AddressZero } from "@ethersproject/constants";
+import { getSafeWithSingleton, migrationContractTo150, getSafeSingletonAt, getMock } from "../utils/setup";
+import deploymentData from "../json/safeDeployment.json";
+import safeRuntimeBytecode from "../json/safeRuntimeBytecode.json";
+import { executeContractCallWithSigners } from "../../src/utils/execution";
+
+const SAFE_SINGLETON_150_ADDRESS = "0x88627c8904eCd9DF96A572Ef32A7ff13b199Ed8D";
+
+const SAFE_SINGLETON_150_L2_ADDRESS = "0x0Ee37514644683f7EB9745a5726C722DeBa77e52";
+
+const COMPATIBILITY_FALLBACK_HANDLER_150 = "0x8aa755cB169991fEDC3E306751dCb71964A041c7";
+
+const FALLBACK_HANDLER_STORAGE_SLOT = "0x6c9a6c4a39284e37ed1cf53d337577d14212a4870fb976a4366c693b939918d5";
+
+const GUARD_STORAGE_SLOT = "0x4a204f620c8c5ccdca3fd54d003badd85ba500436a431f0cbda4f558c93c34c8";
+
+describe("Safe150Migration library", () => {
+    const migratedInterface = new ethers.Interface(["function masterCopy() view returns(address)"]);
+
+    const setupTests = deployments.createFixture(async ({ deployments }) => {
+        await deployments.fixture();
+
+        // Set the runtime code for hardcoded addresses, so the expected events are emitted
+        await hre.network.provider.send("hardhat_setCode", [SAFE_SINGLETON_150_ADDRESS, safeRuntimeBytecode.safe150]);
+        await hre.network.provider.send("hardhat_setCode", [SAFE_SINGLETON_150_L2_ADDRESS, safeRuntimeBytecode.safe150l2]);
+        await hre.network.provider.send("hardhat_setCode", [
+            COMPATIBILITY_FALLBACK_HANDLER_150,
+            safeRuntimeBytecode.safe150CompatibilityFallbackHandler,
+        ]);
+
+        const signers = await ethers.getSigners();
+        const [user1] = signers;
+        const singleton130Address = (await (await user1.sendTransaction({ data: deploymentData.safe130 })).wait())?.contractAddress;
+        const singleton130L2Address = (await (await user1.sendTransaction({ data: deploymentData.safe130l2 })).wait())?.contractAddress;
+
+        if (!singleton130Address || !singleton130L2Address) {
+            throw new Error("Could not deploy Safe130 or Safe130L2");
+        }
+        const singleton130 = await getSafeSingletonAt(singleton130Address);
+        const singleton130L2 = await getSafeSingletonAt(singleton130L2Address);
+
+        const guardContract = await hre.ethers.getContractAt("Guard", AddressZero);
+        const guardEip165Calldata = guardContract.interface.encodeFunctionData("supportsInterface", ["0x945b8148"]);
+        const validGuardMock = await getMock();
+        await validGuardMock.givenCalldataReturnBool(guardEip165Calldata, true);
+
+        const invalidGuardMock = await getMock();
+        await invalidGuardMock.givenCalldataReturnBool(guardEip165Calldata, false);
+
+        const migration = await (await migrationContractTo150()).deploy();
```

Comment:
Why don't we deploy the migration contract as part of the deployment scripts and then reused the deterministic instance of it?

Reply:
I thought about doing that, but I couldn't come up with a neat way to do this because it depends on the singleton addresses, which either have to be computed depending on the runtime bytecode (doesn't work for the future contract updates well) or hardcoding the addresses (they're not known yet, because the 1.5.0 contract is not released). I plan on adding it to the deploy scripts with hardcoded addresses alongside the 1.5.0 release.